### PR TITLE
CRM-21197 Convert from html to plain text AFTER token substitution has happened

### PR DESCRIPTION
**Pull request by michaelmcandrew**
Overview
----------------------------------------
Convert any html contained in tokens to plain text for plain text emails
--

Before
----------------------------------------
HTML in tokens that contained HTML was being outputted to plain text versions of emails.

After
----------------------------------------
Any HTML in tokens in converted to plain text.

Technical Details
----------------------------------------
~~The entire text version of the email is converted again. At first this felt like it might be expensive, but 
it is probably cheaper than checking all of the tokens for HTML.~~

The other approach would be to insert the tokens before the email is converted to plain text but I think there are reasons that this does not happen in the current implemented - I didn't have time to investigate why this happens but I am guessing it is something to do with specific types of tokens.

Comments
----------------------------------------
Since we are just converting something that is suppoosedly plain text into plain text, I can't see why this would be dangerous.

---

 * [CRM-21197: CiviMail does not convert html in tokens to plain text in plain text emails](https://issues.civicrm.org/jira/browse/CRM-21197)

### CODE REVIEW

**Comment by seamuslee001** [comment]
@michaelmcandrew seem to have got a few failures on Jenkins which all appear to be relevant

**Comment by michaelmcandrew** [comment]
Damn those tests. I'll check it out in the morning...

**Comment by michaelmcandrew** [comment]
OK - I pushed some more commits. The change is now more significant but hopefully better thought through, and all the tests passed locally.

Background:

There was an issue with running CRM_Utils_String::htmlToText multiple times on different chunks of the email as it likes to extract links and put them at the end. This was causing the formatting of the links section to get mucked up, and the test suite was quite rightly complaining about this.

After a bit of head stratching, I decided that the only way to avoid this would be to do the conversion from html to plain text (when the plain text version has not been supplied) AFTER token substitution has happened.

The latest PR does this by moving responsibility for the creation of the plain text version of the email from getTemplates to compose().

As a result it is called once per email sent rather than once mailing job which will slow things down (the previous PR did this as well) but I doubt the performance hit will be massive as all CRM_Utils_String::htmlToText does is call rcube_html2text which is a fairly small library that is just doing regex search and replace and no DB or disk IO.

**Comment by michaelmcandrew** [comment]
@seamuslee001 - i'm confused as to why the checks are failing - they are passing locally - do we need to start them again?

**Comment by seamuslee001** [comment]
Jenkins re test this please

**Comment by seamuslee001** [comment]
@michaelmcandrew I think Tim restarted the ubuntu server

**Comment by michaelmcandrew** [comment]
Thanks @seamuslee001.

From the test results, the only failure was "Failure in api call for CxnApp get:  CRL distributor has an invalid certificate", which I am assuming is unrelated.

**Comment by seamuslee001** [comment]
yep looks about right

**Comment by seamuslee001** [comment]
Jenkins re test this please

**Comment by michaelmcandrew** [comment]
@seamuslee001 can we merge this now?

**Comment by mlutfy** [comment]
This seems good to me (running html2text after tokens have been generated). If there are no objections, I will merge by the end of the day.

**Comment by michaelmcandrew** [comment]
great, thanks @mlutfy

**Comment by mlutfy** [comment]
hmm, I tested this patch but it does not seem to work as intended:

capture d ecran de 2017-11-17 16-10-13

Steps to reproduce:

* I created a contact custom field of type "Note" with WYSIWYG.
* Set my mailing preferences to "Text only"
* Added two lines of text, including one with bold.
* Sent myself a test mailing.

Result: the text version included HTML.

**Comment by mlutfy** [comment]
Even with the `$test/text` typo fixed, it still doesn't work for me. While debugging, it seems like the `$pEmails['text']` is already populated by the time it gets to this check line ~ 1306:

```
      if ($text) {
        // If the text version exists, use it
        $textBody = implode('', $text);
      }
      elseif ($html) {
        // Else if it doesn't exist and the html version exists, use it
        $textBody = implode('', $html);
        $textBody = CRM_Utils_String::htmlToText($textBody);
      }
```

This seems unrelated to your patch, but from the CiviMail UI, it shows the text part as empty:

capture d ecran de 2017-11-17 16-23-35

**Comment by michaelmcandrew** [comment]
>  from the CiviMail UI, it shows the text part as empty:
That is as expected.

I'll try your steps to reproduce and see what happens - thanks for getting involved :)

**Comment by michaelmcandrew** [comment]
@mlutfy Thanks again for testing :) I looked into this some more.

It turns out that the patch I wrote only worked when there was no email header. This was still an improvement on the status quo (and worked for my use case as I wasn't using headers) but I dug a bit more to find out what was happening.

It turns out that there was another(!) place in the code that was creating a plain text version of the email when one didn't exist: getTemplates().

The logic in that function went along the lines of: If there is a text version of the email OR a plain text header, then return the plain text version of the email (with the header and footer appended) for further processing.

You were using a header. That is why $text was populated by that point.

That logic seems wrong to me. We should only return a plain text version of the email if we have written one. If we care about whether the header exists, we should also care about the footer. But I don't think we should care about either.

The condition was imported like that from svn days so not sure why it exists (and don't think there is an easy way to check (correct me if I am wrong)).

I have taken out the extra condition and as a result $text is not populated at the time that you would check and the whole email is converted at that time from html to plain text.

PS. You mentioned a $test/$text typo. I don't think that is a typo. $test is passed in as a param and used in one other (similar) place in that function. I'm not sure how it is used but I don't think the changes I have made to the code will affect anything in that regard.

**Comment by michaelmcandrew** [comment]
I'll check those three failing tests and let you know...

**Comment by michaelmcandrew** [comment]
I've made a few more changes to get these tests to pass:

Automatically generated plain text versions of emails have a different (hopefully you will agree improved) layout in some cases. Hence some of the regex needed re-ordering.

I used the s modifier in the regexes which so we didn't have to put as many `\n`s into the tests (though some `\n`s are significant and I left these in.

Raw URLs (i.e. not defined with an <a> tag) are not tracked in html emails. The test suite suggested that they would be tracked when converted to plain text. In my testing, this did not happen. I suspect because of changes in the order in which stuff is converted. I adjusted the test to account for this> I think it is a reasonable behaviour since in this instance, the test version is a backup/fallback, and it is strange that they should be tracked in one format and not in another.

The sample HTML header and footer now have some `<p>` tags in them. This improves rendering when they are converted to plain text (it inserts new lines) which makes the tests more realistic. It required a change to `xml/templates/civicrm_data.tpl` but I think this will have a minimal impact on real life usage since no one actually uses the sample header and footer. I didn't include `sql/civicrm_generated.mysql` - not sure on the protocol there.

**Comment by michaelmcandrew** [comment]
The test suite reports a failure `CRM_Activity_ActionMappingTest::testDefault.testDefault with data set #3`. I can't reproduce this locally.

Any thoughts?

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by michaelmcandrew** [comment]
@eileenmcnaughton @mlutfy - all tests passed - can we merge?!

**Comment by mlutfy** [comment]
(*[CiviCRM Review Template](https://docs.civicrm.org/dev/en/latest/standards/review/) WORD-1.0*)

* ([`r-jira`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-jira)) ✔ PR title/description is clear
* ([`r-test`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-test)) ✔ they pass
* ([`r-code`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-code)) ✔ looks clear
* ([`r-doc`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-doc)) ✔ n/a
* ([`r-maint`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-maint)) ✔ I think so
* ([`r-run`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-run)) ✔
  * I tested: create a contact custom field, insert that field as a token in a mailing (`{contact.custom_7}`), then send the email (with logging to DB enabled). I checked in the DB, and the plain + html versions were giving the expected result.
  * I wasn't too sure what else to test (what is affected). I tested Contact > Send Email. It worked as expected.
* ([`r-user`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-user)) ✔ if users liked html in plain text emails, that's kind of weird. I suspect users are OK with this change.
* ([`r-tech`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-tech)) ✔ I don't think there is a technical impact, other than improving CiviCRM.

File: tests/phpunit/CRM/Mailing/BaseMailingSystemTest.php [hunk 1]
```diff
@@ -315,11 +307,10 @@ public function urlTrackingExamples() {
       array('url_tracking' => 1),
     );
     $cases[] = array(
-      // Plain-text URL's are tracked in plain-text emails...
-      // but not in HTML emails.
+      // Plain-text URLs are not tracked.
       "<p>Please go to: http://example.net/</p>",
       ";<p>Please go to: http://example\.net/</p>;",
-      ';Please go to: .*extern/url.php\?u=\d+&qid=\d+;',
+      ';Please go to: http://example\.net/;',
```
Comment (totten) [inline tests/phpunit/CRM/Mailing/BaseMailingSystemTest.php]: This seems like a significant policy change...

**Comment by totten** [comment]
This seems to have caused some regressions in FlexMailer:

 * https://test.civicrm.org/job/CiviCRM-Ext-Matrix/CIVIVER=master,EXTKEY=org.civicrm.flexmailer,label=ubuntu1604/109/#showFailuresLink
 * https://github.com/civicrm/org.civicrm.flexmailer/blob/master/tests/phpunit/Civi/FlexMailer/FlexMailerSystemTest.php

I haven't had a chance to fully digest this PR yet, but at a high-level... I agree there's an issue where the substance of a token should be different for plain-text and HTML. A good example might be a token that embeds a small table reporting about your recent donations. The plain-text and HTML versions should look different.

In `TokenProcessor` (which is part of core -- and used for FlexMailer), the ambiguity was resolved with a more robust API for defining tokens ([here](https://docs.civicrm.org/dev/en/latest/framework/civimail/#extending-the-token-system)). This allows one to:

 * Define token in plain-text: `$row->format('text/plain')->tokens(...)`. (Let the system autoconvert to HTML.)
 * Define token in HTML: `$row->format('text/html')->tokens(...)`. (Let the system autoconvert to text.)
 * Define both plain-text and HTML.

File: tests/phpunit/CRM/Mailing/BaseMailingSystemTest.php (hunk 1 above)
Comment (totten) [reply tests/phpunit/CRM/Mailing/BaseMailingSystemTest.php]: Mixed feelings about disabling click-tracking on plain-text mailings:
 * On text mailings, you'll probably get more recipients clicking the link if the URL looks sensible. I'd guess that a majority of users would prefer to enable click-tracking on HTML but disable click-tracking on text.
 * I don't think *all* users would, and I think the change is a little big for a point-release. I'd rather see some kind of toggle to allow smoother transition.
 * I'm very much open to improving FlexMailer to make this easier to manipulate programmatically. (It should actually be possible already by modifying or replacing the `'civi_flexmailer_text_click_tracker`, but it's a little obscure -- the [docs](https://docs.civicrm.org/flexmailer/en/latest/develop/#services) mention the service but describe how to replace it. That's something like `$container->findDefinition('civi_flexmailer_text_click_tracker')->setMethodCalls(...'myfunc'...)`. I'd be make a prettier DX.)

**Comment by michaelmcandrew** [comment]
@totten - back from the christmas break today - I'll take a look at this and respond soon.

**Pull request merged**